secp256k1-sys/src/lib.rs

```diff
@@ -569,6 +569,33 @@ pub unsafe fn secp256k1_context_destroy(ctx: *mut Context) {
     rustsecp256k1_v0_4_1_context_destroy(ctx)
 }
 
+/// FFI-safe replacement for panic
+///
+/// Prints to stderr and aborts with `std`, double panics without `std`.
+#[cfg_attr(not(feature = "std"), allow(unused))]
+fn ffi_abort(msg: impl core::fmt::Display) -> ! {
+    #[cfg(feature = "std")]
+    {
+        eprintln!("[libsecp256k1] {}", msg);
+        std::process::abort()
+    }
+    #[cfg(not(feature = "std"))]
+    {
+        use core::fmt::Display;
+
+        // Abort by double panic
+        struct PanicOnDrop<M: Display>(M);
+
+        impl<T: Display> Drop for PanicOnDrop<T> {
+            fn drop(&mut self) {
+                panic!("[libsecp256k1] {}", self.0);
+            }
+        }
+
+        let _bomb = PanicOnDrop(&msg);
+        panic!("[libsecp256k1] {}", &msg)
```

Comment:
I think this is still UB. The unwinder works in two passes. In the first pass it looks for a catch landing pad. In the second pass it actually unwinds. The double panic only happens on the second pass, but on the first pass I think optimizations due to the nounwind attribute may already cause UB. One optimization I can imagine is that LLVM sees that for example `rustsecp256k1_v0_4_1_default_illegal_callback_fn` can't unwind and then propagates this info to `ffi_abort` which then r

Reply:
That makes sense, great that you pointed it out!

I will wrap it in `catch_unwind` then and put `loop {}` at the end just in case.

Reply:
That should indeed fix the UB I think.

Reply:
Unfortunately I just found that `catch_unwind` is unavailable without `std`, looks like the only way is to loop forever. :(

Reply:
Right, forgot about that. Does this library ever get used in places where there is no libc or would it be possible to call `abort()` from libc? I believe that is what `std::process::abort()` does on most platforms.

Reply:
Or maybe write a C function that uses a compiler intrinsic to abort and then call this function?

Reply:
Yeah, using C function is probably the best way, especially since this crate already contains lot of C.

Reply:
@bjorn3 Yes, this library is widely used in wasm32-unknown-unknown
We thought about writing a small c function that calls `__builtin_trap` but that precludes MSVC: https://github.com/rust-bitcoin/rust-secp256k1/pull/288#issuecomment-799537632

This brings me back to this: https://github.com/rust-bitcoin/rust-secp256k1/issues/354#issuecomment-1005710448

Reply:
(side note, Thank you for taking the time to look at this and respond here! @bjorn3 )

Reply:
I think we should let it go.